tendermint/src/lite/verifier.rs

```diff
@@ -154,3 +165,137 @@ where
 
     Ok(())
 }
+
+/// Returns Ok if we can trust the passed in (untrusted) header
+/// based on the given trusted state, otherwise returns an Error.
+fn can_trust<TS, SH, VS, L, S, R>(
+    trusted_state: &TS,    // h1 in spec
+    untrusted_header: &SH, // h2 in spec
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = VS>,
+    SH: SignedHeader,
+    VS: ValidatorSet,
+    L: TrustThreshold,
+    S: Store<SignedHeader = SH>,
+    R: Requester<SignedHeader = SH, ValidatorSet = VS>,
+{
+    // can we trust the still untrusted header based on the given trusted state?
+    match check_support(
+        trusted_state,
+        untrusted_header,
+        trust_threshold,
+        trusting_period,
+        now,
+    ) {
+        Ok(_) => {
+            store.add(untrusted_header)?;
+            return Ok(());
+        }
+        Err(e) => {
+            if e != Error::InsufficientVotingPower {
+                return Err(e);
```

Comment:
Not sure this is right. I think it's a bug in the spec too. If we do this, won't we never actually enter bisection?

Reply:
Not sure I'm following. I think the above logic is a bit difficult to read and might be confusing. We can probably do better! 

But as far as I understood, bisection starts below and only in the case `InsufficientVotingPower` (called `ErrTooMuchChange` in spec and go-code).

We could change `check_support` to return a `Result<bool, Error>` instead of now `Result<(), Error>`. In the case of `TooMuchChange` it would return false, indicating we do not have enough support. In every other case we return an error. 

In any case, we are trying to deal with 3 different scenarios here: 
- enough support: store the new header to trusted store and we are done
- something went wrong: some real error occurred which could be just bubbled up
- there was not enough support: no real error but we need to do more to verify if we can trust that header ...

Reply:
Yep my bad I thought we were always returning there but it's clear that we pass through if the err is InsuficientVotingPower. I wonder if it could be made more clear eg with a match statement?

Reply:
#96